crates/wasmtime/src/runtime/vm/libcalls.rs

```diff
@@ -829,13 +848,16 @@ unsafe fn array_new_elem(
 
         let array = match ArrayRef::_new_fixed(store, &pre, &vals) {
             Ok(a) => a,
-            Err(e) if e.is::<GcHeapOutOfMemory<()>>() => {
-                // Collect garbage to hopefully free up space, then try the
-                // allocation again.
-                store.maybe_async_gc(None)?;
-                ArrayRef::_new_fixed(store, &pre, &vals)?
+            Err(e) => {
+                if let Some(oom) = e.downcast_ref::<crate::GcHeapOutOfMemory<()>>() {
```

Comment:
Similar to tests, this pattern is 3 times in this file, would it be possible to refactor that to share code amongst the users?